File: VatApp_Net/XmlGenerator.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Xml;
using VatApp_Net.Models;

namespace VatApp_Net
{
    public class XmlGenerator
    {
        public string GenerateZip(DeclarationTva decla, string outputPath)
        {
            string tempXmlPath = Path.Combine(Path.GetTempPath(), $"tva_{DateTime.Now:yyyyMMddHHmmss}.xml");
            
            var settings = new XmlWriterSettings { Indent = true, OmitXmlDeclaration = false };
            using (var writer = XmlWriter.Create(tempXmlPath, settings))
            {
                writer.WriteStartDocument(true);
                writer.WriteStartElement("releveDeductions");
                
                writer.WriteElementString("identifiantFiscal", decla.IdentifiantFiscal);
                writer.WriteElementString("ice", decla.ICE);
                writer.WriteElementString("annee", decla.Annee.ToString());
                writer.WriteElementString("periode", decla.Periode.ToString());
                writer.WriteElementString("regime", decla.Regime.ToString());
                
                writer.WriteStartElement("rdps");
                foreach (var d in decla.Deductions)
                {
                    writer.WriteStartElement("rdp");
                    writer.WriteElementString("ord", d.Ordre.ToString());
                    writer.WriteElementString("num", d.NumeroFacture);
                    writer.WriteElementString("des", d.NomFournisseur);
                    writer.WriteElementString("tva", d.MontantTva.ToString("F2").Replace(",", "."));
                    writer.WriteElementString("ht", d.MontantHT.ToString("F2").Replace(",", "."));
                    writer.WriteElementString("ttc", d.MontantTTC.ToString("F2").Replace(",", "."));
                    writer.WriteElementString("ref", d.ReferenceEcriture);
                    writer.WriteElementString("tx", d.TauxTva.ToString("0").Replace(",", "."));
                    writer.WriteElementString("mp", d.ModePaiement.ToString());
                    writer.WriteElementString("dp", d.DatePaiement.ToString("yyyy-MM-dd"));
                    writer.WriteElementString("df", d.DateFacture.ToString("yyyy-MM-dd"));
                    writer.WriteElementString("if", d.IF_Fournisseur);
                    writer.WriteElementString("ice", d.ICE_Fournisseur);
                    writer.WriteEndElement(); // rdp
                }
                writer.WriteEndElement(); // rdps
                writer.WriteEndElement(); // releveDeductions
            }

            if (File.Exists(outputPath)) File.Delete(outputPath);

            using (ZipArchive zip = ZipFile.Open(outputPath, ZipArchiveMode.Create))
            {
                zip.CreateEntryFromFile(tempXmlPath, Path.GetFileName(tempXmlPath));
            }

            if (File.Exists(tempXmlPath)) File.Delete(tempXmlPath);
            
            return outputPath;
        }
    }
}

File: SageToVatXml/XmlGenerator.vb
Imports System.Xml.Linq
Imports System.IO
Imports System.IO.Compression
Imports SageToVatXml.Models

Public Class XmlGenerator
    Public Sub GenerateZip(declaration As DeclarationTva, outputPath As String)
        Dim xmlDoc = New XDocument(
            New XDeclaration("1.0", "UTF-8", "yes"),
            New XElement("releveDeductions",
                New XElement("identifiantFiscal", declaration.IdentifiantFiscal),
                New XElement("ice", declaration.ICE),
                New XElement("annee", declaration.Annee),
                New XElement("periode", declaration.Periode),
                New XElement("regime", declaration.Regime),
                New XElement("rdps",
                    declaration.Deductions.Select(Function(d, index)
                        Return New XElement("rdp",
                            New XElement("ord", index + 1),
                            New XElement("num", d.NumeroFacture),
                            New XElement("des", d.NomFournisseur),
                            New XElement("tva", d.MontantTva.ToString("F2").Replace(",", ".")),
                            New XElement("ht", d.MontantHT.ToString("F2").Replace(",", ".")),
                            New XElement("ttc", d.MontantTTC.ToString("F2").Replace(",", ".")),
                            New XElement("ref", d.NumeroFacture),
                            New XElement("tx", d.TauxTva),
                            New XElement("mp", d.ModePaiement),
                            New XElement("dp", d.DatePaiement.ToString("yyyy-MM-dd")),
                            New XElement("df", d.DateFacture.ToString("yyyy-MM-dd")),
                            New XElement("if", d.IdFournisseur),
                            New XElement("ice", d.ICEFournisseur)
                        )
                    End Function)
                )
            )
        )

        Dim tempXmlPath = Path.Combine(Path.GetTempPath(), $"tva_{DateTime.Now:yyyyMMddHHmmss}.xml")
        xmlDoc.Save(tempXmlPath)

        If File.Exists(outputPath) Then File.Delete(outputPath)

        Using zip As ZipArchive = ZipFile.Open(outputPath, ZipArchiveMode.Create)
            ' Le fichier à l'intérieur du ZIP doit souvent avoir un nom spécifique ou simplement .xml
            zip.CreateEntryFromFile(tempXmlPath, Path.GetFileName(tempXmlPath).Replace(".xml", ".xml"))
        End Using

        File.Delete(tempXmlPath)
    End Sub
End Class
